extensions/sockets/src/main/java/org/mule/module/socket/api/client/UdpClient.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.socket.api.client;
+
+import static java.util.Arrays.copyOf;
+import static org.mule.module.socket.internal.SocketUtils.createPacket;
+import static org.mule.module.socket.internal.SocketUtils.getAddress;
+import static org.mule.module.socket.internal.SocketUtils.getUdpAllowedByteArray;
+import org.mule.module.socket.api.exceptions.ReadingTimeoutException;
+import org.mule.module.socket.api.socket.udp.DefaultUdpClientSocketProperties;
+import org.mule.module.socket.api.source.ImmutableSocketAttributes;
+import org.mule.module.socket.api.source.SocketAttributes;
+import org.mule.runtime.core.api.serialization.ObjectSerializer;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.SocketAddress;
+import java.net.SocketTimeoutException;
+
+/**
+ * This {@link SocketClient} implementation allows the reading and writing
+ * to and from a specific UDP {@link DatagramSocket}.
+ */
+public final class UdpClient implements SocketClient
+{
+
+    private final DefaultUdpClientSocketProperties socketProperties;
+    private final ObjectSerializer objectSerializer;
+    private final DatagramSocket socket;
+    private final SocketAddress socketAddress;
+
+    public UdpClient(DatagramSocket socket, String host, int port, DefaultUdpClientSocketProperties socketProperties, ObjectSerializer objectSerializer)
+    {
+        this.objectSerializer = objectSerializer;
+        this.socketProperties = socketProperties;
+        this.socket = socket;
+        this.socketAddress = getAddress(host, port);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void write(Object data) throws IOException
+    {
+        byte[] byteArray = getUdpAllowedByteArray(data, objectSerializer);
+        DatagramPacket sendPacket = createPacket(byteArray);
+        sendPacket.setSocketAddress(socketAddress);
+        socket.send(sendPacket);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public InputStream read() throws IOException
+    {
+        DatagramPacket receivedPacket = createPacket(socketProperties.getReceiveBufferSize());
+        receivedPacket.setSocketAddress(socketAddress);
+        try
+        {
+            socket.receive(receivedPacket);
+            return new ByteArrayInputStream(copyOf(receivedPacket.getData(), receivedPacket.getLength()));
```

Comment:
I did that just for the sake of shrinking the byte[] buffer to the actual packet length. If not, I would have a InputStream larger than needed and full of padding.